src/posix.c

```diff
@@ -216,9 +217,44 @@ int p_write(git_file fd, const void *buf, size_t cnt)
 	return 0;
 }
 
-#ifdef NO_MMAP
+ssize_t p_pread(git_file fd, void *buf, size_t nbyte, git_off_t offset)
+{
+#ifdef GIT_WIN32
+	/* FIXME: Implement using Win32 overlapped IO */
 
-#include "map.h"
+	ssize_t ret;
+	int nfd = dup(fd);
+	
+	(void) ((ret = p_lseek(nfd, offset, SEEK_SET) < 0 ? -1 : 0) ||
+		((ret = p_read(nfd, buf, nbyte)) < 0));
+
+	close(nfd);
+	return ret;
+#else
+	char *b = (char *)buf;
+	while (nbyte) {
+		ssize_t r;
+		r = pread(fd, b, nbyte, offset);
+		if (r < 0) {
+			if (errno == EINTR || GIT_ISBLOCKED(errno))
+				continue;
+			return -1;
+		}
+		if (!r) {
+			errno = EPIPE;
```

Comment:
`EPIPE` is not in the list of errors that can be returned by either [`read(2)`](https://man7.org/linux/man-pages/man2/read.2.html#ERRORS) or [`lseek(2)`](https://man7.org/linux/man-pages/man2/lseek.2.html#ERRORS) (which [`pread(2)`](https://man7.org/linux/man-pages/man2/pread.2.html#ERRORS) claims to be the legal errors).

instead, maybe break here like in the `p_read()` implementation. the caller would still need to check whether the return value was a short read, for consistency with `p_read()`, but that's already done.